File: apps/universal_recognizer_web/core/accessibility.py
"""
Accessibility detection and suggestions for universal character recognition.
"""

from typing import Dict, List, Optional


# Common dyslexia confusion pairs
DYSLEXIA_CONFUSION_PAIRS = {
    'b': 'd',
    'd': 'b',
    'p': 'q',
    'q': 'p',
    'n': 'u',
    'u': 'n',
    'm': 'w',
    'w': 'm',
    '6': '9',
    '9': '6',
    'I': 'l',
    'l': 'I',
    'O': '0',
    '0': 'O'
}


def detect_potential_issues(prediction_result: Dict, mirror_result: Optional[Dict] = None, 
                           quality_metrics: Optional[Dict] = None) -> List[Dict]:
    """
    Detect potential accessibility issues from prediction results.
    
    Args:
        prediction_result: Standard prediction result
        mirror_result: Mirror detection result (if available)
        quality_metrics: Writing quality metrics (if available)
    
    Returns:
        List of detected issues with severity and type
    """
    issues = []
    
    if prediction_result is None:
        return issues
    
    predicted_char = prediction_result.get('predicted_character', '')
    confidence = prediction_result.get('confidence', 0.0)
    
    # Low confidence issue
    if confidence < 60.0:
        issues.append({
            'type': 'low_confidence',
            'severity': 'medium',
            'message': f'Low confidence ({confidence:.1f}%) - try drawing more clearly',
            'suggestion': 'Make your strokes bolder and keep the character centered'
        })
    elif confidence < 40.0:
        issues.append({
            'type': 'very_low_confidence',
            'severity': 'high',
            'message': f'Very low confidence ({confidence:.1f}%) - character may be unclear',
            'suggestion': 'Try redrawing the character with clearer, bolder strokes'
        })
    
    # Mirror detection issue
    if mirror_result and mirror_result.get('mirror_detected', False):
        mirrored_char = mirror_result.get('predicted_character', '')
        issues.append({
            'type': 'mirror_detected',
            'severity': 'high',
            'message': f'Mirror detected: Did you mean to write "{mirrored_char}"?',
            'suggestion': f'Your character appears mirrored. Try writing "{mirrored_char}" in the correct orientation.',
            'original_char': predicted_char,
            'mirrored_char': mirrored_char
        })
    
    # Dyslexia confusion pattern detection
    if predicted_char in DYSLEXIA_CONFUSION_PAIRS:
        confused_char = DYSLEXIA_CONFUSION_PAIRS[predicted_char]
        
        # Check if confused character is in top predictions
        top_predictions = prediction_result.get('top_predictions', [])
        confused_in_top = any(
            pred.get('character') == confused_char 
            for pred in top_predictions[:3]  # Check top 3
        )
        
        if confused_in_top:
            issues.append({
                'type': 'dyslexia_confusion',
                'severity': 'medium',
                'message': f'Common confusion: "{predicted_char}" and "{confused_char}" look similar',
                'suggestion': f'Make sure you wrote "{predicted_char}" and not "{confused_char}". Pay attention to the direction.',
                'confused_pair': (predicted_char, confused_char)
            })
    
    # Writing quality issues
    if quality_metrics:
        overall_score = quality_metrics.get('overall_score', 100.0)
        clarity_score = quality_metrics.get('clarity_score', 100.0)
        centering_score = quality_metrics.get('centering_score', 100.0)
        size_score = quality_metrics.get('size_score', 100.0)
        
        if overall_score < 50.0:
            issues.append({
                'type': 'poor_quality',
                'severity': 'medium',
                'message': 'Writing quality could be improved',
                'suggestion': 'Try making your strokes clearer and keeping the character well-centered'
            })
        
        if clarity_score < 40.0:
            issues.append({
                'type': 'low_clarity',
                'severity': 'low',
                'message': 'Strokes could be bolder',
                'suggestion': 'Press harder or use a thicker brush to make your strokes more visible'
            })
        
        if centering_score < 50.0:
            issues.append({
                'type': 'poor_centering',
                'severity': 'low',
                'message': 'Character could be better centered',
                'suggestion': 'Try to draw the character in the center of the canvas'
            })
        
        if size_score < 30.0:
            issues.append({
                'type': 'too_small',
                'severity': 'low',
                'message': 'Character is quite small',
                'suggestion': 'Try drawing a larger character that fills more of the canvas'
            })
        elif size_score > 90.0:
            issues.append({
                'type': 'too_large',
                'severity': 'low',
                'message': 'Character is very large',
                'suggestion': 'Try drawing a slightly smaller character so it fits better'
            })
    
    return issues


def generate_suggestions(issues: List[Dict], predicted_char: str) -> List[Dict]:
    """
    Generate helpful suggestions based on detected issues.
    
    Args:
        issues: List of detected issues
        predicted_char: The predicted character
    
    Returns:
        List of suggestions with actionable advice
    """
    suggestions = []
    
    # Sort issues by severity
    severity_order = {'high': 0, 'medium': 1, 'low': 2}
    sorted_issues = sorted(issues, key=lambda x: severity_order.get(x.get('severity', 'low'), 2))
    
    for issue in sorted_issues:
        suggestion = {
            'type': issue.get('type'),
            'priority': issue.get('severity', 'low'),
            'message': issue.get('message', ''),
            'advice': issue.get('suggestion', ''),
            'category': _categorize_issue(issue.get('type'))
        }
        suggestions.append(suggestion)
    
    # Add character-specific guidance if no major issues
    if len(issues) == 0 or all(i.get('severity') == 'low' for i in issues):
        char_guidance = _get_character_guidance(predicted_char)
        if char_guidance:
            suggestions.append({
                'type': 'character_guidance',
                'priority': 'info',
                'message': f'Writing tip for "{predicted_char}"',
                'advice': char_guidance,
                'category': 'guidance'
            })
    
    return suggestions


def _categorize_issue(issue_type: str) -> str:
    """Categorize issue type."""
    if issue_type in ['mirror_detected', 'dyslexia_confusion']:
        return 'accessibility'
    elif issue_type in ['low_confidence', 'very_low_confidence', 'poor_quality']:
        return 'quality'
    elif issue_type in ['low_clarity', 'poor_centering', 'too_small', 'too_large']:
        return 'technique'
    else:
        return 'general'


def _get_character_guidance(char: str) -> Optional[str]:
    """Get character-specific writing guidance."""
    guidance = {
        'b': 'Start from the top, draw a straight line down, then add the curve on the right',
        'd': 'Start with the curve on the left, then add the straight line',
        'p': 'Draw the line down first, then add the curve at the bottom',
        'q': 'Draw the curve first, then add the line going down',
        '6': 'Start from the top, curve down and around',
        '9': 'Start from the bottom, curve up and around',
        '0': 'Draw a complete circle or oval',
        'O': 'Similar to zero, but make sure it\'s clearly a letter',
        'I': 'A single vertical line - make it straight',
        'l': 'Similar to I, but make sure it\'s lowercase',
        'a': 'Start with the circle, then add the line on the right',
        'e': 'Start with the horizontal line, then add the curve',
        's': 'Draw an S-curve, making sure both curves are clear'
    }
    return guidance.get(char.lower())


def get_resources(issue_type: str) -> List[Dict]:
    """
    Get educational resources based on issue type.
    
    Args:
        issue_type: Type of issue detected
    
    Returns:
        List of resources with links and descriptions
    """
    resources = []
    
    if issue_type in ['mirror_detected', 'dyslexia_confusion']:
        resources.extend([
            {
                'title': 'Understanding Letter Reversals',
                'description': 'Learn about why letter reversals happen and how to practice',
                'category': 'dyslexia',
                'url': 'https://www.understood.org/en/articles/why-kids-reverse-letters'
            },
            {
                'title': 'Handwriting Practice for Kids',
                'description': 'Free printable worksheets for practicing letter formation',
                'category': 'practice',
                'url': 'https://www.education.com/worksheets/handwriting/'
            }
        ])
    
    if issue_type in ['low_confidence', 'poor_quality']:
        resources.extend([
            {
                'title': 'Improving Handwriting Skills',
                'description': 'Tips and exercises for clearer handwriting',
                'category': 'technique',
                'url': 'https://www.handwritingforkids.com/'
            }
        ])
    
    # General resources
    resources.extend([
        {
            'title': 'NeuralEngine Documentation',
            'description': 'Learn more about how character recognition works',
            'category': 'general',
            'url': '#'
        }
    ])
    
    return resources


def format_accessibility_report(prediction_result: Dict, mirror_result: Optional[Dict] = None,
                               quality_metrics: Optional[Dict] = None) -> Dict:
    """
    Create a comprehensive accessibility report.
    
    Args:
        prediction_result: Standard prediction result
        mirror_result: Mirror detection result
        quality_metrics: Writing quality metrics
    
    Returns:
        Complete accessibility report
    """
    issues = detect_potential_issues(prediction_result, mirror_result, quality_metrics)
    suggestions = generate_suggestions(issues, prediction_result.get('predicted_character', ''))
    
    # Get resources for all issue types
    all_resources = []
    for issue in issues:
        resources = get_resources(issue.get('type', ''))
        all_resources.extend(resources)
    
    # Remove duplicates
    seen_urls = set()
    unique_resources = []
    for resource in all_resources:
        if resource['url'] not in seen_urls:
            seen_urls.add(resource['url'])
            unique_resources.append(resource)
    
    return {
        'issues': issues,
        'suggestions': suggestions,
        'resources': unique_resources,
        'has_issues': len(issues) > 0,
        'severity_summary': {
            'high': sum(1 for i in issues if i.get('severity') == 'high'),
            'medium': sum(1 for i in issues if i.get('severity') == 'medium'),
            'low': sum(1 for i in issues if i.get('severity') == 'low')
        }
    }

File: apps/universal_recognizer_web/static/js/accessibility.js
// NeuralEngine Universal Character Recognizer - Accessibility Features
window.updateAccessibilityDisplay = function(result) {
  if (!result) {
    clearAccessibilityDisplay();
    return;
  }

  updateMirrorDetection(result.mirror_detection);
  updateWritingQuality(result.quality_metrics);
  updateSuggestions(result.accessibility);
  updateResources(result.accessibility);
}

function clearAccessibilityDisplay() {
  document.getElementById("mirrorStatus").innerHTML = '<p>Draw a character to analyze...</p>';
  document.getElementById("qualityMetrics").innerHTML = '<p>Analyze your writing...</p>';
  document.getElementById("suggestionsList").innerHTML = '<p>No suggestions yet</p>';
  
  // Always show resources from the start
  updateResourcesInitial();
}

function updateMirrorDetection(mirrorData) {
  const container = document.getElementById("mirrorStatus");
  
  if (!mirrorData || !mirrorData.mirror_detected) {
    container.innerHTML = `
      <div class="mirror-status mirror-not-detected">
        No mirror writing detected. Character orientation looks correct!
      </div>
    `;
    return;
  }

  const original = mirrorData.original_prediction;
  const mirrored = mirrorData.mirrored_prediction;
  
  container.innerHTML = `
    <div class="mirror-status mirror-detected">
      <strong>Mirror Detected!</strong>
      <p>Your character appears to be mirrored.</p>
      <p>Original: "${original.predicted_character}" (${original.confidence.toFixed(1)}% confidence)</p>
      <p>Mirrored: "${mirrored.predicted_character}" (${mirrored.confidence.toFixed(1)}% confidence)</p>
      <p>Did you mean to write "${mirrored.predicted_character}"?</p>
    </div>
  `;
}

function updateWritingQuality(qualityMetrics) {
  const container = document.getElementById("qualityMetrics");
  
  if (!qualityMetrics) {
    container.innerHTML = '<p>Quality analysis unavailable</p>';
    return;
  }

  const scores = [
    { label: 'Overall', value: qualityMetrics.overall_score },
    { label: 'Clarity', value: qualityMetrics.clarity_score },
    { label: 'Size', value: qualityMetrics.size_score },
    { label: 'Centering', value: qualityMetrics.centering_score },
    { label: 'Stroke', value: qualityMetrics.stroke_score }
  ];

  container.innerHTML = '<div class="quality-metrics">' +
    scores.map(score => {
      let qualityClass = 'quality-low';
      if (score.value >= 90) qualityClass = 'quality-high';
      else if (score.value >= 75) qualityClass = 'quality-medium';
      
      return `
        <div class="quality-score">
          <div class="quality-label">
            <span>${score.label}:</span>
            <span class="quality-value">${score.value.toFixed(0)}%</span>
          </div>
          <div class="quality-bar">
            <div class="quality-bar-fill ${qualityClass}" style="width: ${score.value}%"></div>
          </div>
        </div>
      `;
    }).join('') +
    '</div>';
}

function updateSuggestions(accessibility) {
  const container = document.getElementById("suggestionsList");
  
  if (!accessibility || !accessibility.suggestions || accessibility.suggestions.length === 0) {
    container.innerHTML = `
      <div class="suggestions-list">
        <div class="suggestion-item" style="border-color: rgba(34, 197, 94, 0.5); background: rgba(34, 197, 94, 0.05);">
          <div class="suggestion-message" style="color: #22c55e;">✓ No issues detected - great writing!</div>
        </div>
      </div>
    `;
    return;
  }

  container.innerHTML = '<div class="suggestions-list">' +
    accessibility.suggestions.map(suggestion => {
      const priorityClass = suggestion.priority === 'high' ? 'high-priority' : 
                            suggestion.priority === 'medium' ? 'medium-priority' : '';
      
      return `
        <div class="suggestion-item ${priorityClass}">
          <div class="suggestion-message">${suggestion.message}</div>
          ${suggestion.advice ? `<div class="suggestion-advice">${suggestion.advice}</div>` : ''}
        </div>
      `;
    }).join('') +
    '</div>';
}

function updateResourcesInitial() {
  const container = document.getElementById("resourcesList");
  
  // Default resources shown from the start
  const defaultResources = [
    {
      title: "Neural Engine Documentation",
      description: "Learn more about how character recognition works and explore the Neural Engine framework.",
      url: "https://github.com/matt-minev/NeuralEngine"
    },
    {
      title: "Improve Your Handwriting",
      description: "Tips and techniques for clearer character formation and better recognition accuracy.",
      url: "https://www.wikihow.com/Improve-Your-Handwriting"
    }
  ];
  
  container.innerHTML = defaultResources.map(resource => `
    <div class="resource-item">
      <div class="resource-title">${resource.title}</div>
      <div class="resource-description">${resource.description}</div>
      <a href="${resource.url}" target="_blank" class="resource-link">Learn more →</a>
    </div>
  `).join('');
}

function updateResources(accessibility) {
  const container = document.getElementById("resourcesList");
  
  if (!accessibility || !accessibility.resources || accessibility.resources.length === 0) {
    // If no custom resources, show defaults
    updateResourcesInitial();
    return;
  }

  container.innerHTML = accessibility.resources.map(resource => {
    const url = resource.url === '#' ? '#' : resource.url;
    return `
      <div class="resource-item">
        <div class="resource-title">${resource.title}</div>
        <div class="resource-description">${resource.description}</div>
        ${url !== '#' ? `<a href="${url}" target="_blank" class="resource-link">Learn more →</a>` : ''}
      </div>
    `;
  }).join('');
}

// Initialize accessibility display on page load
document.addEventListener("DOMContentLoaded", () => {
  clearAccessibilityDisplay();
});

